File: app/src/main/java/com/rakangsoftware/architecture/screen/contacts/ContactsViewModel.java
package com.rakangsoftware.architecture.screen.contacts;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.support.annotation.NonNull;

import com.rakangsoftware.architecture.data.AppDatabase;
import com.rakangsoftware.architecture.data.Contact;
import com.rakangsoftware.architecture.data.ContactRepository;
import com.rakangsoftware.architecture.utils.SingleLiveEvent;

import java.util.List;


public class ContactsViewModel extends AndroidViewModel {

    private final ContactRepository mContactRepository;

    private final SingleLiveEvent<Boolean> mShowDialog     = new SingleLiveEvent<>();
    private final SingleLiveEvent<Contact> mDeletedContact = new SingleLiveEvent<>();
    private Contact mDelContact;

    public ContactsViewModel(@NonNull final Application application) {
        super(application);

        mContactRepository = new ContactRepository(
                AppDatabase.getDB(application).getContactDao()
        );
    }

    public LiveData<List<Contact>> getContacts() {
        return mContactRepository.getAll();
    }

    public LiveData<Boolean> getShowDialog() {
        return mShowDialog;
    }

    public SingleLiveEvent<Contact> getDeletedContact() {
        return mDeletedContact;
    }

    public void onContactClicked(final Contact contact) {
        // Open edit!!!
    }

    public void onDeleteContact(final Contact contact) {
        mDelContact = contact;
        mContactRepository.delete(contact);
        mDeletedContact.setValue(contact);
    }

    public void doDelete(final Contact contact) {
        mDelContact = null;
    }

    public void restoreContact(final Contact contact) {
        mContactRepository.create(contact);
    }

    public void clickAddContact() {
        mShowDialog.postValue(true);
    }

    public void createContact(final Contact contact) {
        mContactRepository.create(contact);
    }


}

File: app/src/main/java/com/rakangsoftware/architecture/screen/contacts/ContactsViewModel.kt
package com.rakangsoftware.architecture.screen.contacts

import android.app.Application
import android.arch.lifecycle.AndroidViewModel
import android.arch.lifecycle.LiveData

import com.rakangsoftware.architecture.data.AppDatabase
import com.rakangsoftware.architecture.data.Contact
import com.rakangsoftware.architecture.data.ContactRepository
import com.rakangsoftware.architecture.utils.SingleLiveEvent


class ContactsViewModel(application: Application) : AndroidViewModel(application) {

    private val mContactRepository: ContactRepository

    private val mShowDialog = SingleLiveEvent<Boolean>()
    val deletedContact = SingleLiveEvent<Contact>()
    private var mDelContact: Contact? = null

    val contacts: LiveData<List<Contact>>
        get() = mContactRepository.all

    val showDialog: LiveData<Boolean>
        get() = mShowDialog

    init {

        mContactRepository = ContactRepository(
                AppDatabase.getDB(application).contactDao
        )
    }

    fun onContactClicked(contact: Contact) {
        // Open edit!!!
    }

    fun onDeleteContact(contact: Contact) {
        mDelContact = contact
        mContactRepository.delete(contact)
        deletedContact.value = contact
    }

    fun doDelete(contact: Contact) {
        mDelContact = null
    }

    fun restoreContact(contact: Contact) {
        mContactRepository.create(contact)
    }

    fun clickAddContact() {
        mShowDialog.postValue(true)
    }

    fun createContact(contact: Contact) {
        mContactRepository.create(contact)
    }


}
